### DISCUSSION TITLE
Warnings about instance replacement.

### DISCUSSION THREAD

**Discussion by michaelbrewer** [aws/aws-cdk#16057] [General]
# The Idea?

Warn users when there will be an instance replacement on deployment, and potentially an option to block an deployment when this could happen.

Even in the CFN spec json, there is an indication of the `UpdateType` being mutable (update with some interruption or none) vs immutable, being a full instance replacement.

```json5
"Type": {
  "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-apigateway-documentationpart-location.html#cfn-apigateway-documentationpart-location-type",
  "PrimitiveType": "String",
  "Required": false,
  "UpdateType": "Immutable"
}
```

- Docs on  [Update behavior](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-updating-stacks-update-behaviors.html)

### Update with No Interruption
AWS CloudFormation updates the resource without disrupting operation of that resource and without changing the resource's physical ID. For example, if you update certain properties on an AWS::CloudTrail::Trail resource, AWS CloudFormation updates the trail without disruption.

### Updates with Some Interruption
AWS CloudFormation updates the resource with some interruption. For example, if you update certain properties on an AWS::EC2::Instance resource, the instance might have some interruption while AWS CloudFormation and Amazon EC2 reconfigure the instance.

### Replacement
AWS CloudFormation recreates the resource during an update, which also generates a new physical ID. AWS CloudFormation usually creates the replacement resource first, changes references from other dependent resources to point to the replacement resource, and then deletes the old resource. For example, if you update the AvailabilityZone property of an AWS::EC2::Instance resource type, AWS CloudFormation creates a new resource and replaces the current EC2 Instance resource with the new one.

**Comment by peterwoodworth** [comment]
I like this idea @michaelbrewer 🙂

Currently, you can use `cdk diff` to inspect the differences in your template. You could certainly write a script to check for resource replacement yourself, but I think lots of users could benefit from a toggleable safety net preventing deployment on resource replacement.

Even a warning sounds like it could be useful, I know lots of CDK users can get tripped up by the way CloudFormation handles resource replacement

**Reply by peterwoodworth** [reply]
I've created a feature request so that we can track this change. Contributions are welcome as well 🙂 

https://github.com/aws/aws-cdk/issues/19838

**Reply by michaelbrewer** [reply]
I was thinking of some kind of switch to list out whether replacements or downtime was expected.
